README.rst
==============
Jupyter-Invoke
==============

This package provides an extension to Jupyter notebook server, which enables
executing notebooks non-interactively and obtaining JSON or CSV output with
HTTP requests.

Installation
============

.. code-block:: bash

    pip install git+https://github.com/EniramLtd/jupyter-invoke.git

You can enable the server extension in all new notebook servers by
running

.. code-block:: bash

    jupyter serverextension enable --py jupyter_invoke

You can also enable the extension at command line when you a start
particular server:

.. code-block:: bash

    jupyter notebook --NotebookApp.nbserver_extensions="{'jupyter_invoke':True}"

Alternatively, you can set the variable in a `Jupyter config file`_:

.. code-block:: python

    c = get_config()
    c.NotebookApp.nbserver_extensions = {
        'jupyter_invoke': True,
    }

.. _`Jupyter config file`: http://jupyter-notebook.readthedocs.io/en/latest/config_overview.html

Usage
=====

In a Jupyter notebook, call :func:`jupyter_invoke.response.respond` at the end
of one cell to mark it as providing the invocation output:

.. code-block:: python

    from jupyter_invoke import respond
    a = {'a': [1, 2, 3]}
    respond(a)

This will mark the cell as providing JSON output when querying
``http://<server_address>:<port>/invoke/<path/to/notebook.ipynb>``


Allowing only certain notebooks to be invoked
---------------------------------------------

It is possible to restrict the invocation to only particularly
named notebooks.  There is a configuration parameters
``Invoker.filename_pattern`` which, if specified, matches the end of
the notebook filename (before ``.ipynb``) before allowing the invocation.
Non-matching notebooks are then invisible to the invoke handler.  As an
example, consider the following jupyter configuration file that allows only
files with pattern ``*_invoke.ipynb`` to be invoked:

.. code-block:: python

    c = get_config()
    c.Invoker.filename_pattern = '_invoke'

By default, all notebooks can be invoked.


Using query parameters in invocation
------------------------------------

Parameters can be passed to the notebook in URL query strings by using
:func:`jupyter_invoke.response.get_param`.  For instance,

.. code-block:: python

    from jupyter_invoke import get_param, respond
    key = get_param('key', 'a', optional=False)
    a = {key: [1, 2, 3]}
    respond(a)

Now we would need to invoke the notebook with a query string like
``?key=b`` or otherwise the there will be an error.  In interactive
use the notebook will use the second, default value, parameter for
the variable.  If the ``optional`` flag is set to True, the default
value may also be used in invocation if the parameters isn't provided
in the query.


Error cases
===========

If the requested notebook cannot be found, the HTML status
code will be 404.  In other errors the HTML status code may
be 200 (OK), but the output is a JSON string that contains
the field ``error`` and an explanation.


Limitations
===========

Only Python 3 and notebook versions above 4 are supported.
``python setup.py test`` can be used for testing but it
does not cover the actual integration with a server.



CHANGES.rst
Change log
==========


0.3.0 / 2017-05-10
------------------
- Feature: Support query parameters in invocation URL. (#3)


0.2.0 / 2017-04-27
~~~~~~~~~~~~~~~~~~
- Feature: Enable limiting invokable notebooks based on their names. (#2)



jupyter_invoke/tests/test_invoke.py
from jupyter_invoke.invoke import execute_notebook, InvokeException
import pytest
import os

@pytest.mark.parametrize('nb_name,expected,status_code,error_extra',[
    ('notebooks/success_csv.ipynb', (""";a;b
2017-01-01;1;3
2017-01-02;2;4
2017-01-03;3;5\n""", 'text/csv'), None, None),
    ('notebooks/success_json.ipynb',
     ({"a": {"2017-01-01T00:00:00.000Z": 1,
             "2017-01-02T00:00:00.000Z": 2},
       "b": {"2017-01-01T00:00:00.000Z": 3,
             "2017-01-02T00:00:00.000Z": 4}}, 'application/json'), None, None),
    ('notebooks/no_output_cell.ipynb', 'Cannot find a cell', 200, None),
    ('notebooks/many_output_cells.ipynb', 'Multiple cells with', 200,
     [({'a': 1, 'b': 2}, 'application/json'), ({'a': 1}, 'application/json')])
])
def test_notebooks(nb_name, expected, status_code, error_extra):
    os.chdir(os.path.dirname(__file__))
    try:
        result = execute_notebook(nb_name)
        assert result == expected
    except InvokeException as e:
        assert expected in str(e)
        assert e.status_code == status_code
        assert expected in e.error_dict['error']
        if error_extra:
            assert error_extra == e.error_dict['outputs']


def test_notebook_query_parameters():
    os.chdir(os.path.dirname(__file__))
    params = {'par1': [b'10'], 'par3': [b'30']}
    result = execute_notebook('notebooks/query_params.ipynb', params)
    assert result == ({'params': ['10', 2, '30']}, 'application/json')


jupyter_invoke/respond.py
"""Wrappers for Jupyter cell output for parsing the invocation output

The wrappers provide different forms for the output so that
they are displayed nicely in a notebook in a browser and so
that the invocation finds the right piece of data.

Much of the wrapper functionality concerns Pandas DataFrames.  If not
working with Pandas DataFrames, JSON output is works with
``dict`` too.

Use :func:`respond` to wrap the cell output for invocation.

"""
from IPython.display import JSON
import json

try:
    import pandas as pd
except ImportError:
    pd = None

INVOKE_TAG = 'invoke_output'

CSV_OUTPUT = 'text/csv'
JSON_OUTPUT = 'application/json'


def _is_dataframe(x):
    return pd and isinstance(x, pd.DataFrame)


class RespondJSON(JSON):
    def __init__(self, x, **kwargs):
        super(RespondJSON, self).__init__(x)
        self.json_params = kwargs

    def _check_data(self):
        if _is_dataframe(self.data):
            return
        super(RespondJSON, self)._check_data()

    def _repr_html_(self):
        if _is_dataframe(self.data):
            return self._data.to_html()

    def _repr_json_(self):
        """Dict and metadata dict"""
        if _is_dataframe(self.data):
            return (json.loads(self.data.to_json(**self.json_params)),
                    {INVOKE_TAG: JSON_OUTPUT})
        return self.data, {INVOKE_TAG: JSON_OUTPUT}

    def __repr__(self):
        return repr(self.data)


class RespondCSV(RespondJSON):
    """Wrap the CSV output inside the JSON field

    The CSV is put into the JSON field and in the notebook
    execution phase it is extracted from it.  This work-around
    was made because IPython did not have built-in support for CSV.
    
    """
    def __init__(self, x, **kwargs):
        super(RespondCSV, self).__init__(x)
        self.csv_params = kwargs

    def _check_data(self):
        if not _is_dataframe(self.data):
            raise ValueError('Can make CSV output only from Pandas DataFrames')

    def _repr_json_(self):
        """Dict and metadata dict"""
        return ({CSV_OUTPUT: self.data.to_csv(**self.csv_params)},
                {INVOKE_TAG: JSON_OUTPUT})


def respond(output, mimetype=JSON_OUTPUT, **kwargs):
    """Wrap Jupyter cell output for invocation

    ``mimetype`` should be "application/json" or "text/csv".
    However, the latter is only supported with Pandas DataFrames.

    ``kwargs`` are passed to Pandas DataFrame ``to_json``
    or ``to_csv`` method if ``output`` is a DataFrame.

    Examples:

    At the end of the notebook cell, call ``respond``:

    >>> a = {'a': [1, 2], 'b': [3, 4]}
    >>> respond(a, JSON_OUTPUT)
    {'a': [1, 2], 'b': [3, 4]}

    Now the notebook cell output is marked as the output
    for the invocation and this is provided as JSON.

    """
    if mimetype == JSON_OUTPUT:
        return RespondJSON(output, **kwargs)
    elif mimetype == CSV_OUTPUT:
        return RespondCSV(output, **kwargs)
    else:
        raise ValueError('Unknown output type {}, use "{}" or "{}".'
                         ''.format(mimetype, JSON_OUTPUT, CSV_OUTPUT))


jupyter_invoke/__init__.py
from traitlets.config import LoggingConfigurable
from traitlets import Unicode
from notebook.utils import url_path_join
from .invoke import InvokeNotebookHandler
from .response import respond, JSON_OUTPUT, CSV_OUTPUT, get_param


class Invoker(LoggingConfigurable):
    """Configuration parameters for the extensions"""
    filename_pattern = Unicode(
        help="""
        Regex for matching the end of notebook name (before ``.ipynb``)
        to allow invocation.  If not specified, all notebooks are allowed
        to be invoked.
        """
    ).tag(config=True)


def _get_config(web_app):
    """Parsing of the configuration parameters"""
    filename_pattern = web_app.settings['config'].get(
        'Invoker', {}).get('filename_pattern') or ""

    return {'filename_pattern': filename_pattern}


def _jupyter_server_extension_paths():
    return [{
        "module": "jupyter_invoke"
    }]


def load_jupyter_server_extension(nb_server_app):
    """Called when the extension is loaded

    Adapted from
    http://jupyter-notebook.readthedocs.io/en/latest/extending/handlers.html

    """
    web_app = nb_server_app.web_app
    filename_pattern = _get_config(web_app)['filename_pattern']
    host_pattern = '.*$'
    route_pattern = url_path_join(web_app.settings['base_url'],
                                  '/invoke/(?P<notebook_name>.*{}\.ipynb)'
                                  '(?:\?.*)?$'
                                  ''.format(filename_pattern))
    web_app.add_handlers(host_pattern, [(route_pattern,
                                         InvokeNotebookHandler)])


jupyter_invoke/response.py
"""Wrappers for Jupyter cell output for parsing the invocation output

The wrappers provide different forms for the output so that
they are displayed nicely in a notebook in a browser and so
that the invocation finds the right piece of data.

Much of the wrapper functionality concerns Pandas DataFrames.  If not
working with Pandas DataFrames, JSON output is works with
``dict`` too.

Use :func:`respond` to wrap the cell output for invocation.

"""
from IPython.display import display, JSON, Javascript
import json

try:
    import pandas as pd
except ImportError:
    pd = None

INVOKE_TAG = 'invoke_output'

CSV_OUTPUT = 'text/csv'
JSON_OUTPUT = 'application/json'

# This may be assigned to use query parameters with get_param function
QUERY_PARAMETERS = None


def _is_dataframe(x):
    return pd and isinstance(x, pd.DataFrame)


def get_param(name, default, optional=False):
    """Pass query parameters in notebook invocation
    
    This function can be used to use HTTP query parameters
    in an invoked notebook.  In interactive use this uses the
    given default value.
    
    From query parameters this always takes the first occurrence
    of the corresponding parameter and returns it as a string.
    
    By default, this ignores the default value when used in
    invocation and raises an exception if the parameters is not
    provided in the query.
    
    """
    if QUERY_PARAMETERS is None:
        return default
    if name in QUERY_PARAMETERS:
        return QUERY_PARAMETERS[name][0].decode()
    elif optional:
        return default
    else:
        raise ValueError('Required parameter "{}" missing from the query '
                         'string'.format(name))


class RespondJSON(JSON):
    def __init__(self, x, **kwargs):
        super(RespondJSON, self).__init__(x)
        self.json_params = kwargs

    def _check_data(self):
        if _is_dataframe(self.data):
            return
        super(RespondJSON, self)._check_data()

    def _repr_html_(self):
        if _is_dataframe(self.data):
            return self._data.to_html()

    def _repr_json_(self):
        """Dict and metadata dict"""
        if _is_dataframe(self.data):
            return (json.loads(self.data.to_json(**self.json_params)),
                    {INVOKE_TAG: JSON_OUTPUT})
        return self.data, {INVOKE_TAG: JSON_OUTPUT}

    def __repr__(self):
        return repr(self.data)


class RespondCSV(RespondJSON):
    """Wrap the CSV output inside the JSON field

    The CSV is put into the JSON field and in the notebook
    execution phase it is extracted from it.  This work-around
    was made because IPython did not have built-in support for CSV.

    """
    def __init__(self, x, **kwargs):
        super(RespondCSV, self).__init__(x)
        self.csv_params = kwargs

    def _check_data(self):
        if not _is_dataframe(self.data):
            raise ValueError('Can make CSV output only from Pandas DataFrames')

    def _repr_json_(self):
        """Dict and metadata dict"""
        return ({CSV_OUTPUT: self.data.to_csv(**self.csv_params)},
                {INVOKE_TAG: JSON_OUTPUT})


def respond(output, mimetype=JSON_OUTPUT, **kwargs):
    """Wrap Jupyter cell output for invocation

    ``mimetype`` should be "application/json" or "text/csv".
    However, the latter is only supported with Pandas DataFrames.

    ``kwargs`` are passed to Pandas DataFrame ``to_json``
    or ``to_csv`` method if ``output`` is a DataFrame.

    Examples:

    At the end of the notebook cell, call ``respond``:

    >>> a = {'a': [1, 2, 3]}
    >>> respond(a, JSON_OUTPUT)
    <IPython.core.display.Javascript object>
    The following output will be provided as JSON in invocation:
    {'a': [1, 2, 3]}
    >>> import pandas as pd
    >>> b = pd.DataFrame(a)
    >>> b['b'] = [4, 5, 6]
    >>> respond(b, CSV_OUTPUT)
    <IPython.core.display.Javascript object>
    The following output will be provided as CSV in invocation:
       a  b
    0  1  4
    1  2  5
    2  3  6

    In the above example the JavaScript object means that the
    notebook displays the URL for the invocation.

    Now the notebook cell output is marked as the output
    for the invocation from the URL of the notebook where
    ``/notebooks/`` is replaced with ``/invoke/``. 

    """
    output_type = mimetype.split('/')[-1].upper()
    _js_invoke_path(output_type)
    print('The following output will be provided as {} in invocation:'
          ''.format(output_type))
    if mimetype == JSON_OUTPUT:
        return RespondJSON(output, **kwargs)
    elif mimetype == CSV_OUTPUT:
        return RespondCSV(output, **kwargs)
    else:
        raise ValueError('Unknown output type {}, use "{}" or "{}".'
                         ''.format(mimetype, JSON_OUTPUT, CSV_OUTPUT))


def _js_invoke_path(output_type):
    """Display the invoke address in Jupyter"""
    display(Javascript("""
        var address = window.location.href.replace('/notebooks/', '/invoke/');
        alert('{} available at ' + address + ' after saving the notebook ' +
              '(the notebook server may impose limitations regarding ' +
              'acceptable notebook file names).');
        """.format(output_type)))


jupyter_invoke/invoke.py
import nbformat
from nbconvert.preprocessors import ExecutePreprocessor
from nbconvert.preprocessors.execute import CellExecutionError
from notebook.base.handlers import IPythonHandler

import os

from .response import CSV_OUTPUT, JSON_OUTPUT, INVOKE_TAG


class InvokeNotebookHandler(IPythonHandler):
    """Execute notebook and return the output"""
    def get(self, notebook_name):
        try:
            query_params = self.request.arguments
            output, mimetype = execute_notebook(notebook_name, query_params)
            output_filename = os.path.basename(notebook_name)
            output_filename = output_filename.replace('.ipynb', '')
            ext = mimetype.split('/')[-1]
            self.set_header('Content-Type', mimetype)
            self.set_header('Content-Disposition',
                            'inline; filename="{}.{}"'
                            ''.format(output_filename, ext))
            self.finish(output)
        except InvokeException as e:
            self.set_status(e.status_code)
            self.write(e.error_dict)


class InvokeException(Exception):
    """Expected error from execution or parsing of output"""
    def __init__(self, msg, results=None, status_code=200, *args, **kwargs):
        super(InvokeException, self).__init__(msg, *args, **kwargs)
        if results is None:
            self.error_dict = {'error': msg}
        else:
            self.error_dict = {'error': msg, 'outputs': results}
        self.status_code = status_code


def handle_query_parameters(notebook, query_parameters):
    """Insert a cell in the beginning of the notebook for query parameters
    
    This modifies the notebook so that :func:`jupyter_invoke.respond.get_param`
    look for keys and values from given HTTP request parameters
    
    """
    first_cell = {'cell_type': 'code',
                  'source': "from jupyter_invoke import response\n"
                             "response.QUERY_PARAMETERS = {!r}"
                             "".format(query_parameters)}
    notebook['cells'].insert(0, nbformat.notebooknode.NotebookNode(first_cell))


def execute_notebook(notebook_name, query_params=None):
    """Execute the notebook from path and parse the output"""
    if not os.path.exists(notebook_name):
        msg = 'Cannot find notebook {}'.format(notebook_name)
        raise InvokeException(msg, status_code=404)
    with open(notebook_name) as f:
        notebook = nbformat.read(f, as_version=4)
    handle_query_parameters(notebook, query_params)
    ep = ExecutePreprocessor(timeout=300, kernel_name='python3')
    try:
        executed_notebook = ep.preprocess(notebook,
                                          {'metadata': {'path': ''}})
    except CellExecutionError as e:
        msg = 'Unable to execute notebook {}\n{}'.format(notebook_name,
                                                         str(e))
        raise InvokeException(msg)

    return parse_execute_output(executed_notebook, notebook_name)


def parse_execute_output(executed_notebook, notebook_name):
    """Parse the output cells for 'invoke_output'

    The output metadata of one cell should define 'invoke_output' key
    (it should be inside "application/json" field). This key indicates
    which field is used for the output dict.

    CSV output is extracted from inside the "application/json" field
    of the output (see the FIXME below).

    The cell output should be created with
    :func:`jupyter_invoke.respond.respond`.

    """
    results = []
    for nb in executed_notebook:
        for cell in nb.get('cells', []):
            for output in cell.get('outputs', []):
                if (output['output_type'] == 'execute_result'
                        and JSON_OUTPUT in output['metadata']
                        and INVOKE_TAG in output['metadata'][JSON_OUTPUT]):
                    output_type = output['metadata'][JSON_OUTPUT][INVOKE_TAG]
                    result = output['data'][output_type]
                    # FIXME: Ugly special case for CSV output
                    if CSV_OUTPUT in result:
                        result = result[CSV_OUTPUT]
                        output_type = CSV_OUTPUT
                    results.append((result, output_type))

    if len(results) == 0:
        msg = ('Cannot find a cell with a proper output metadata tag "{}" '
               'from notebook {}'.format(INVOKE_TAG, notebook_name))
        raise InvokeException(msg)
    elif len(results) > 1:
        msg = ('Multiple cells with output metadata tag "{}" in {}'
               ''.format(INVOKE_TAG, notebook_name))
        raise InvokeException(msg, results)

    return results[0]
